add letterbox-like resize function in the python api

LWQ2EDU opened the pull request:
Implement a letterbox-like resize function for images with an aspect ratio that exceeds the specified ratio in the Python API. The function can be controlled to enable or disable by the parameter `use_letterbox_like`, which is set to enable (True) by default.
![test image](https://github.com/RapidAI/RapidOCR/assets/117698078/de950e7d-7a8b-40f9-9bc7-bd75a349a82f)
![OCR result for test image](https://github.com/RapidAI/RapidOCR/assets/117698078/6e110a9b-ccde-419d-bb87-4b10f06b6adc)
![letterbox-like image](https://github.com/RapidAI/RapidOCR/assets/117698078/b58f40c7-d348-418e-9bd4-2963d020c47f)
![OCR result for letterbox-like image](https://github.com/RapidAI/RapidOCR/assets/117698078/eb009283-10c7-4adb-8729-3b52cff1b038)

LWQ2EDU commented:
![20240202130927](https://github.com/RapidAI/RapidOCR/assets/117698078/c054dd39-4a98-4c3a-996d-55b852cad41b)

SWHL commented:
感谢，我晚上看一下。

SWHL commented:
There is a problem with this idea: after adding a letterbox to the image, the subsequent detection result coordinates cannot correspond to the original image.

LWQ2EDU commented:
> There is a problem with this idea: after adding a letterbox to the image, the subsequent detection result coordinates cannot correspond to the original image.
> 
> Is it convenient for you to make changes here?

Fixed by adding a parameter to record the difference between the resized and original image

SWHL commented:
LGTM

SWHL commented:
@LWQ2EDU 
I made some logical optimizations to the adding letterbox part, removed the use_letterbox_like parameter, and changed it to trigger add letterbox when the aspect ratio and min_height are not met.

I hope you can help review it to see if there are any flaws.

https://github.com/RapidAI/RapidOCR/blob/c2b12e51fbe9737ae54d5514395268272b70e166/python/rapidocr_onnxruntime/main.py#L108-L123

LWQ2EDU commented:
> Hi @LWQ2EDU I made some logical optimizations to the adding letterbox part, removed the use_letterbox_like parameter, and changed it to trigger add letterbox when the aspect ratio and min_height are not met.
> 
> I hope you can help review it to see if there are any flaws.

Using `new_h = w` in L117 may increase the run time of the function. I think change it to `new_h =  max(int(w / self.width_height_ratio), self.min_height) + 1` will be better. In my testing, it is 10 times faster (0.0012s -> 0.0001s).


https://github.com/RapidAI/RapidOCR/blob/c2b12e51fbe9737ae54d5514395268272b70e166/python/rapidocr_onnxruntime/main.py#L117

SWHL commented:
I also notice that. From a time perspective, your approach is indeed faster.
But I test the following image and found that the padding value in your method was too small, so that it could not be recognized correctly.

![](https://raw.githubusercontent.com/RapidAI/RapidOCR/main/python/tests/test_files/test_without_det.jpg)

LWQ2EDU commented:
> I also notice that. From a time perspective, your approach is indeed faster. But I test the following image and found that the padding value in your method was too small, so that it could not be recognized correctly.
> 
> ![](https://raw.githubusercontent.com/RapidAI/RapidOCR/main/python/tests/test_files/test_without_det.jpg)
> 
> Can we compromise and find a balanced strategy?

I think the issue is caused by not optimal `width_height_ratio` and `min_height`. these thresholds are supposed to make sure all images can be recognized correctly when they reach the thresholds. 

Unfortunately, I do not have sufficient data to test the optimal threshold. For the time being, I believe using an image size that is moderately sized may temporarily address this issue.

`new_h = max(int(w / self.width_height_ratio), self.min_height)*2` currently seems to be a compromise.

SWHL commented:
I agree with you that the default value is as you have it.
I will follow up with a blog post about this parameter.
🤝
